clients/pkg/promtail/targets/file/decompresser.go

```diff
@@ -0,0 +1,285 @@
+		running:   atomic.NewBool(false),
+		posquit:   make(chan struct{}),
+		posdone:   make(chan struct{}),
+		done:      make(chan struct{}),
+		position:  pos,
+		decoder:   decoder,
+	}
+
+	go decompresser.readLines()
+	go decompresser.updatePosition()
+	metrics.filesActive.Add(1.)
+	return decompresser, nil
+}
+
+// mountReader instantiate a reader ready to be used by the decompresser.
+//
+// The selected reader implementation is based on the extension of the given file name.
+// It'll error if the extension isn't supported.
+func mountReader(f *os.File, logger log.Logger) (reader io.Reader, err error) {
+	ext := filepath.Ext(f.Name())
+	var decompressLib string
+
+	if strings.Contains(ext, "gz") { // .gz, .tar.gz
+		decompressLib = "compress/gzip"
+		reader, err = gzip.NewReader(f)
+	} else if ext == ".z" {
+		decompressLib = "compress/zlib"
+		reader, err = zlib.NewReader(f)
+	} else if ext == ".zip" {
+		decompressLib = "compress/flate"
+		reader = flate.NewReader(f)
+	} else if ext == ".bz2" {
+		decompressLib = "bzip2"
+		reader = bzip2.NewReader(f)
+	}
+
+	level.Info(logger).Log("msg", fmt.Sprintf("using %q to decompress file %q", decompressLib, f.Name()))
+
+	if reader != nil {
+		return reader, nil
+	}
+
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+
+	return nil, fmt.Errorf("file %q with unsupported extension", f.Name())
+}
+
+func (t *decompresser) updatePosition() {
+	positionSyncPeriod := t.positions.SyncPeriod()
+	positionWait := time.NewTicker(positionSyncPeriod)
+	defer func() {
+		positionWait.Stop()
+		level.Info(t.logger).Log("msg", "position timer: exited", "path", t.path)
+		close(t.posdone)
+	}()
+
+	for {
+		select {
+		case <-positionWait.C:
+			if err := t.MarkPositionAndSize(); err != nil {
+				level.Error(t.logger).Log("msg", "position timer: error getting position and/or size, stopping decompresser", "path", t.path, "error", err)
+				return
+			}
+		case <-t.posquit:
+			return
+		}
+	}
+}
+
+// readLines read all existing lines of the given compressed file.
+//
+// It first decompress the file as a whole using a reader and then it will iterate
+// over its chunks, separated by '\n'.
+// During each iteration, the parsed and decoded log line is then sent to the API with the current timestamp.
+func (t *decompresser) readLines() {
+	level.Info(t.logger).Log("msg", "read lines routine: started", "path", t.path)
+	t.running.Store(true)
+
+	defer func() {
+		t.cleanupMetrics()
+		level.Info(t.logger).Log("msg", "read lines routine finished", "path", t.path)
+		close(t.done)
+	}()
+	entries := t.handler.Chan()
+
+	f, err := os.Open(t.path)
+	if err != nil {
+		level.Error(t.logger).Log("msg", "error reading file", "path", t.path, "error", err)
+		return
+	}
+	defer f.Close()
+
+	r, err := mountReader(f, t.logger)
+	if err != nil {
+		level.Error(t.logger).Log("msg", "error mounting new reader", "err", err)
+		return
+	}
+
+	level.Info(t.logger).Log("msg", "successfully decompressed file", "path", t.path)
```

Comment:
The file was not successfully decompressed at this point. You rather "mounted" a reader onto the file handle.

Reply:
Thanks for spotting this, addressed by https://github.com/grafana/loki/pull/6708/commits/58b5634d729b311bdbba63c7000320fcaf3ce272. I also added the extension to the log message.